teps/0082-workspace-hinting.md

```diff
@@ -0,0 +1,303 @@
+---
+status: proposed
+title: Workspace Hinting
+creation-date: '2021-09-03'
+last-updated: '2021-09-03'
+authors:
+- '@sbwsg'
+---
+
+# TEP-0082: Workspace Hinting
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Workspaces allow Task authors to declare portions of their Task's filesystem to be supplied at runtime by TaskRuns or PipelineRuns. For example a Task may accept a credential via an optional Workspace and a TaskRun might supply it from a Secret. Another Task might write source code to a Workspace and a PipelineRun could bind a Persistent Volume to it so the source can be passed to other PipelineTasks.
+
+Rephrasing this slightly: the interface that Workspaces expose caters to a number of pretty disjoint use-cases - it's general-purpose. A down-side of that is Task authors can't communicate a Workspace's intended usage in a machine-readable way. There's no way for an author to indicate "this Workspace is intended to accept a credential" or "this Workspace should be supplied with configuration". Similarly for Pipeline authors, there's no way to hint that a Workspace is used to shuttle around between Tasks. They can write a human-readable description as part of the workspace declaration but that's essentially useless to an automated system constructing TaskRuns and PipelineRuns.
+
+The purpose of this TEP is to allow Task and Pipeline authors to "hint" about the intended purpose of a Workspace. The idea is that if authors can mark Workspaces with a purpose then automated systems could be designed to submit reasonable default bindings for them.
+
+## Motivation
+
+### Goals
+
+- Provide a way for Tasks and Pipelines to declare the purpose of Workspaces in a machine-readable format.
+
+### Non-Goals
+
+- Adding constraint-checking or any other logic to Pipelines to validate bound Workspaces based on workspace hints.
```

Comment:
I'm curious why not - even if it wasn't in the initial implementation maybe this would be a reasonable follow on? curious if you excluded this just to keep the scope down or if there's another motivation that it might be interesting to elaborate on

Reply:
maybe it's something like: even if the pipeline author hints that a volume claim could be used, the pipeline might still function correctly with other types and we don't want to arbitrarily prevent that?

Reply:
Yeah you've hit the nail on the head with both your comments here.

If we expand scope to include validation then we'll end up burning a lot of cycles designing the shape of that, how it works, and how configurable it is. I allude to this in the "Questions around future extensions stemming from this change are quite nuanced" bit later on but I'll bring some of that same sentiment up here to support this Non-Goal. I'll also add this as possible future work as well.